src/Passes/Source/AllocationManager/IAllocationManager.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Llvm/Llvm.hpp"
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+    /// Interface class for allocation management. This interface provides means to allocate and release
+    /// statically allocated resources such as qubits and results. In a future version, it may be
+    /// extended with get and store in order to support Arrays and Tuples.
+    class IAllocationManager
+    {
+      public:
+        using Address              = uint64_t;    ///< Value type for address
+        using Index                = uint64_t;    ///< Index type used to access an array element.
```

Comment:
FYI: For my own code I assume that the default type to be used for indexing is `size_t` (which is 32-bit on the 32-bit architecture, 64-bit on the 64-bit architecture, and so on), and there should be a documented or evident reason for using a different type for indexing.

Reply:
I am happy to have a discussion about this. The main reason why I am using fixed with integers here is that if you couple QAT on a 32-bit system (or even 16-bit system) it is not evident that your target system (i.e. the quantum computer) is 16- or 32-bit. That is, the platform used to build the quantum program is not necessarily the execution platform and hence it is not a given that the width should follow the the platform you use to generate the quantum program.

Reply:
I will resolve this, but happy to have an offline discussion around this.

Reply:
OK, sounds reasonable.
(I was just under impression that the `Index` is used inside of a separate tool/executable only (QAT or whatever) running on a PC, and is not included in any way into the generated QIR that is to be executed on the quantum hardware)